### PR TITLE
GH-176 Replace GitHub usernames in body of notifications

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary

This PR adds a function to the template generation process that matches all Github usernames and replaces them with their assigned Mattermost usernames, if any.

The regular expression to match the GitHub usernames is the following:

```
(^|[^_\x60[:alnum:]])(@[[:alnum:]](-?[[:alnum:]]+)*)
\___________________/ \__________/\_______________/
         (1)             (2.1)           (2.2)
                     \_____________________________/
                                    (2)
```

- The first matching group, (1), matches the beginning of the line or a character that is allowed to precede a GitHub username; i.e., any character different than underscore, backtick or an alphanumeric character.
- The second matching group, (2), wraps the matched GitHub username. Notice that there is no limit on its maximum length (39 at the moment), but in the rare case that we would match a string with more than 39 characters, we would not have any mapping to a Mattermost user, so the string will remain unchanged.
    - The (2.1) part makes sure that the username begins with a letter and is at least 1 character long.
    - The (2.2) part allows dashes, if and only if they are immediately followed by a non-empty string of alphanumeric characters. This can be repeated as many times as needed.

The matched strings contain the preceding character, if any. As there is no `ReplaceAllStringFunc` that accepts a function receiving the slice of submatches (see https://github.com/golang/go/issues/5690),  I needed to split the matched string, in the `replaceAllGitHubUsernames` function, into the prefix and the username itself, which we then lookup using the `lookupMattermostUsername` function.

The `replaceAllGitHubUsernames` function is applied to the following templates, when writing the body of the notification:
  - `newPR`
  - `newIssue`
  - `issueComment`
  - `pullRequestReviewEvent`
  - `newReviewComment`
  - `commentMentionNotification`
  - `commentAuthorPullRequestNotification`
  - `pullRequestReviewNotification`

I have also changed the unit tests in `template_test.go`:
- I have adapted the tests for the corresponding templates, adding an additional `"with mentions"` test that contains the usernames in the `usernameMap` keys, that maps GitHub usernames to Mattermost ones. The string with the values of this map is then used as the expected value in the tests. I did it this way so if we need to add an edge case for a GitHub username in the future, we can just modify the map and let the tests themselves unchanged.
- I have also added a unit test for the GitHub regular expression.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-github/issues/176

### CODE REVIEW

File: server/template.go [hunk 1]
```diff
@@ -11,8 +11,10 @@ import (
 )
 
 const mdCommentRegexPattern string = `(<!--[\S\s]+?-->)`
+const gitHubRegexPattern string = `(^|[^_\x60[:alnum:]])(@[[:alnum:]](-?[[:alnum:]]+)*)`
```
Comment (lieut-data) [inline server/template.go]: Thank you for the detailed breakdown in the PR description. Can you also link out to the definition of what constitutes a valid GitHub username, should we need to refer back to that any point?

File: server/template.go [hunk 2]
```diff
@@ -11,8 +11,10 @@ import (
...
 
 var mdCommentRegex = regexp.MustCompile(mdCommentRegexPattern)
+var gitHubRegex = regexp.MustCompile(gitHubRegexPattern)
```
Comment (lieut-data) [inline server/template.go]: should we call this `gitHubUsernameRegex` for clarity?

File: server/template.go (hunk 1 above)
Comment (lieut-data) [reply server/template.go]: I note that this might match `test+@agarciamontoro.com`, which is otherwise a valid email address, but I suspect this would be rare compared to the flexibility of matching `let's ask @lieut-data+@agarciamontoro`, so no changes needed.

File: server/template_test.go [hunk 3]
```diff
@@ -54,11 +69,64 @@ var issue = github.Issue{
...
 	HTMLURL: sToP("https://github.com/panda"),
 }
 
+// A map of known associations between Github users and Mattermost users
+var usernameMap = map[string]string{
+	"panda":          "pandabot",
+	"asaadmahmood":   "asaad.mahmood",
+	"marianunez":     "maria.nunez",
+	"lieut-data":     "jesse.hallam",
+	"sameusername":   "sameusername",
```
Comment (lieut-data) [inline server/template_test.go]: :+1:

File: server/template_test.go [hunk 4]
```diff
@@ -772,6 +1027,64 @@ func TestPullRequestReviewNotification(t *testing.T) {
...
+				State:   sToP("approved"),
+				Body:    sToP("Excited to see git-get-head land!\n" + gitHubMentions),
+			},
+		})
+		require.NoError(t, err)
+		require.Equal(t, expected, actual)
+	}))
+}
+
+func TestGitHubUsernameRegex(t *testing.T) {
```
Comment (lieut-data) [inline server/template_test.go]: 👍

File: server/template.go (hunk 1 above)
Comment (agarciamontoro) [reply server/template.go]: > Can you also link out to the definition of what constitutes a valid GitHub username, should we need to refer back to that any point?

There is no public definition of what constitutes a valid GitHub username. What I have used here are the error messages returned in https://github.com/join (inspired by https://github.com/shinnn/github-username-regex); namely:
  1. The username must be between 1 and 39 characters long.
  2. The username may only contain alphanumeric characters or single hyphens (i.e., not two adjacent hyphens, as in `@wrong--username`).
  3. The username cannot begin or end with a hyphen.

I can add that as a comment before the `gitHubRegexPattern` definition.

> I note that this might match test+@agarciamontoro.com

This matches indeed test+@agarciamontoro, but as you can see in this same comment, GitHub matches it as a username. I think preserving what GitHub does here is the best thing.

File: server/template.go (hunk 2 above)
Comment (agarciamontoro) [reply server/template.go]: Done in https://github.com/mattermost/mattermost-plugin-github/pull/177/commits/5db457eae041ea83f8f438228b523fcd6a600cec.

File: server/template.go (hunk 1 above)
Comment (agarciamontoro) [reply server/template.go]: Done in https://github.com/mattermost/mattermost-plugin-github/pull/177/commits/c43784aa31f7a9352b578750d82c8b4999347dc2.

**Comment by agarciamontoro** [comment]
Just noticed that I was creating the mocked pull request with mentions with the actual Mattermost usernames, not the GitHub ones.
Fixed it in https://github.com/mattermost/mattermost-plugin-github/pull/177/commits/635104fc5d6d85dbc5cdf75d981fba384b73d7aa.

**Comment by DHaussermann** [comment]
@mickmister is there an easy way to get theses templates since I won't test on our production and don't have prod access to initiate these actions so I can't actually get the template
